rust/core-lib/src/editor.rs

```diff
@@ -761,29 +782,36 @@ impl Editor {
     }
 }
 
-#[derive(PartialEq, Eq, Clone, Copy)]
-enum EditType {
+#[derive(PartialEq, Eq, Clone, Copy, Debug, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub enum EditType {
+    /// A catchall for edits that don't fit elsewhere, and which should
+    /// always have their own undo groups; used for things like cut/copy/paste.
     Other,
+    /// An insert from the keyboard/IME (not a paste or a yank).
+    #[serde(rename = "insert")]
     InsertChars,
+    #[serde(rename = "newline")]
+    InsertNewline,
+    /// An indentation adjustment.
+    Indent,
     Delete,
     Undo,
     Redo,
     Transpose,
 }
 
 impl EditType {
-    pub fn json_string(&self) -> &'static str {
-        match *self {
-            EditType::InsertChars => "insert",
-            EditType::Delete => "delete",
-            EditType::Undo => "undo",
-            EditType::Redo => "redo",
-            EditType::Transpose => "transpose",
-            _ => "other",
+    /// Checks whether a new undo group should be created between two edits.
+    fn breaks_undo_group(self, previous: EditType) -> bool {
+        if self == EditType::Other || self == EditType::Transpose {
```

Comment:
I'd personally be inclined to write this `self == EditType::Other || self == EditType::Transpose || self != previous` but it's a style judgment call.